### DISCUSSION TITLE
What format is expected for the innerBlocks array?

### DISCUSSION THREAD

**Discussion by crs1138** [WordPress/gutenberg#38708] [Developer Experience]
First of all, I am probably doing this completely wrong… I just don't know how to approach this better.

I've developed a custom block. It's fairly simple – a wrapper, image and `<InnerBlock />` with a locked template. This all works fine and now, I'd like to create unit tests for this block. My unit tests for other block are using `html-react-parser` processing the output of the `getSaveContent()` and making a snapshot. If anybody refactors a block, snapshots don't match and the dev knows they need to review this.

However, as this block is relying on inner blocks, I am not sure how to go about this. I've noticed that the `getSaveContent()` and `getSaveElement()` functions as defined in the serializer.js take `innerBlocks` as the third argument. The docs say it's an array, but what shape do the individual items/blocks take in that array?

```javascript
getSaveContent(
        {  name, save },
        { ...attributes },
        [ ...innerBlocks ]
);
```

https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/blocks/src/api/serializer.js#L180

I've tried passing in an array of inner blocks as used in `migrate()` in deprecation. I wasn't really expecting that to work, but what if.

Next, I tried to pass the following code:

```javascript
import parse from 'html-react-parser';
import { getSaveContent } from '@wordpress/blocks';
import { save as saveParagraph } from '@wordpress/block-library/src/paragraph/save';
import { default as save } from '../save';

const innerBlocksTemplate = [
    getSaveContent({
        blockDetails: {
            name: 'core/paragraph',
            save: saveParagraph,
        },
        attributes: { content: 'Honza', className: 'cvgb-testimonial__name' },
];

const savedBlock = parser(getSaveContent(
    {
        name: 'cv-wp-gb-lib-blocks/testimonial',
        save,
    },
    { attributes: { faceOption: 'face-1', }  },
    innerBlocksTemplate
));
```

What is a good way to go about this?

**Comment by talldan** [comment]
@crs1138 I don't know `getSaveContent` very well. I'd recommend `serialize` from the `@wordpress/blocks` package as an alternative.

From what I understand it takes an array of block objects. e.g.:
```js
const blockHTML = serialize( [
    {
        name: 'my-block-name',
        attributes: { // ... an object of block attributes },
        innerBlocks: [ // ... an array of inner blocks in the same format as the outer array ],
    },
 ] );
 ```
 The main difference is that I imagine `serialize` will also output the delimiter comment.
 
 Looking at the code, with both `getSaveContent` and `serialize` you'll have to make sure to register the block first in your test using `registerBlockType`. That's because these functions look up the block type definition from an internal registry.
  
This project has some tests that take block html, parse it, apply the block types, and then re-serialize those blocks. They may be useful to look at - https://github.com/WordPress/gutenberg/blob/trunk/test/integration/full-content/full-content.test.js.

**Reply by crs1138** [reply]
Thanks, I'll give it a try.

**Comment by crs1138** [comment]
Perhaps my issue is different after all. Studying the code of the `full-content.test.js`. This is used for generating fixtures. To be honest fixtures and their practicality is still evading me.

We have some custom made Gutenberg blocks that are to be used across a variety of our sites. I want to make sure that the frontend HTML structure of these blocks does not get altered accidentally by any of our developers on the team. Such unplanned alteration could be very difficult to trace across all sites. Am I right to think this is what _fixtures_ are for? Please, correct me if I'm wrong.

**Reply by talldan** [reply]
Yep, that's what the full-content fixtures are used for. Though I think the word 'fixture' is misused a bit here. Often the word 'fixture' means some sample data used as an input for a test. But here it's also used to represent both a fixture and also what's called a snapshot in other parts of the codebase.

A snapshot is some data that a test outputs. The first time a test is run it saves output to a file and the test passes. The second time it runs it compares the current output against the data previously saved to a file and only passes if the two are the same. Snapshots are useful if you want to test that something isn't changing over time.

Those full-content are maybe a little bit more complicated than you might need as they also test some of the internal data structures used by the block editor.

But what they do is the following:
1. Take some HTML (as it might exist in the block editor) for a block from a fixture file
2. Parse it into a general block data structure and compare that data to a snapshot (`wp.blockSerializationDefaultParser.parse`)
3. Parse it applying any known block types, and compare that data to a snapshot (`wp.blocks.parse`)
4. Serialize that block back into HTML and compare that output to a snapshot. (`wp.blocks.serialize`)

It sounds like you only need the parts 1, 3 and 4, and really only needs tests for the output of 4.

**Reply by crs1138** [reply]
Thanks. I'll try it out.

**Reply by crs1138** [reply]
If I may bother you further… 
What do these 3 lines actually do and are they all relevant to my use case (ie if I'm skipping the step 2)?
https://github.com/WordPress/gutenberg/blob/d8ba3977c2da51236bf817eb42f36a077364bc1b/test/integration/full-content/full-content.test.js#L66-L69

**Reply by talldan** [reply]
I think they may be relevant depending on what your blocks do.

The first line loads the `php` code for dynamic blocks. If the blocks that you're testing don't have any php code then you can probably skip this.

The second line is curious (`require( '../../../packages/editor/src/hooks' )`). I think it's supposed to load the code for block supports features. For example, if you have something like `"supports": { "align" // ... }` (or other properties like 'color', 'spacing') in your block.json. But I don't think that will work as those features aren't in that folder any more. It looks like it does nothing. I'll make a PR to remove it.

`registerCoreBlocks` is something you may not need, it registers all the core WordPress blocks (paragraph, quote, image etc.) by calling `wp.blocks.registerBlockType`. This is needed for Gutenberg's test, `parse` and `serialize` won't know how to parse or serialize a paragraph block if it isn't registered. I imagine you're not testing core blocks, so you likely won't need this, but you will need to register your own blocks (using `wp.block.registerBlockType` in a `beforeAll`).

**Reply by crs1138** [reply]
Thank you. We do have a few blocks that rely on ServerSide render so. I will look into the `unstable__bootstrapServerSideBlockDefinitions()` function. However, I shall focus on generating and comparing the fixtures for the JS-only blocks for now.

**Reply by crs1138** [reply]
I've recreated our own `registerCvgbBlocks()` function that is based on Gutenberg's original `registerCoreBlocks()`. I call both of the functions within the `beforeAll()` as some of our custom blocks are using the core blocks as their inner blocks. However, the following error comes back from Jest when I run this:

```javascript
    beforeAll(() => {
        registerCoreBlocks();
        registerCvgbBlocks();
    });
```

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/1313681/166454168-87a64f35-a0d4-4d05-90b9-e22f2122e6d9.png">

The source code for `registerCvgbBlocks()` can be seen here:

```javascript
/**
 * @file The main entry point used to plug in individual Gutenberg blocks and define general settings shared by all of
 *       them.
 * @since 0.1.0
 */
/**
 * WP Gutenberg Block
 * @typedef {object} WPBlock
 * @property {object|undefined} GutenbergBlock - WP Gutenberg Block
 * @property {string} name Block's name
 * @property {Metadata} metadata Block's properties as imported from its `block.json`
 * @property {object} settings Block's properties such as its `save()` or `edit()` functions
 */

/**
 * React Component
 * @typedef {object} ReactComponent
 * @property {object} props
 */

/**
 * WP block metadata
 * @typedef {object} Metadata
 * @property {string} name            Block name in format `namespace/block_name`, ie. `cv-wp-gb-lib-blocks/button`
 * @property {string} title           Block's title
 * @property {string} category        A category the block belongs to
 * @property {object|string} icon     Block's icon
 * @property {string} icon.background Block icon's background
 * @property {string} icon.src        Block icon's file or dashicon name
 * @property {string} description     Block's description
 * @property {array} keywords         Keywords under which block can be found
 * @property {object} attributes      Block's attributes holding the information used in the block
 */

import '@lib/polyfill.min.js';
/**
 * Import registerBlockType blocks
 */

import * as brands from './brands';
import * as breadcrumbs from './breadcrumbs';
import * as card from './card';
import * as cell from './cell';
import * as collapse from './collapse';
import * as compare from './compare';
import * as cvexample from './cv-example';
import * as cvtemplate from './cv-template';
import * as cvword from './cv-word';
import * as grid from './grid';
import * as icons from './icons';
import * as protip from './protip';
import * as protop from './protop';
import * as testimonial from './testimonial';
import * as toclist from './toc-list';

import '@lib/scss/editor/editor.scss';

import(
    /* webpackChunkName: "css/blocks.style" */
    '@lib/scss/front-styles.scss'
);
const { registerBlockType } = wp.blocks;

/**
 * Register a GB Block
 * @param {WPBlock} block
 * @returns undefined
 */
const registerBlock = (block) => {
    if (!block) {
        return;
    }
    const { metadata, settings, name } = block;
    registerBlockType({ name, ...metadata }, settings);
};

/**
 * Get the CVGB blocks we will register
 * @returns Array of blocks to register
 */
export const getCvgbBlocks = () => [
    brands,
    breadcrumbs,
    card,
    cell,
    collapse,
    compare,
    cvexample,
    cvtemplate,
    cvword,
    grid,
    icons,
    protip,
    protop,
    testimonial,
    toclist,
];

export const registerCvgbBlocks = (blocks = getCvgbBlocks()) => {
    blocks.forEach(registerBlock);
};
```

We do employ a `block_categories_all` PHP filter to add our custom category. How could I overcome this issue in Jest environment?

**Reply by talldan** [reply]
I think you can do `wp.blocks.setCategories( [ { slug: 'my-category-slug', title: 'my-category-title' } ] )` in your `beforeAll` to register the category through JavaScript.

One thing to note is that this will unset other categories, including the default ones the block editor provides:https://github.com/WordPress/gutenberg/blob/f8854a2ce3277e4b969246d4bcb1e1fbdb71a4fe/packages/blocks/src/store/reducer.js#L34-L42

If you're only testing blocks that use your custom category, that shouldn't be an issue, but if you're testing blocks that use a default category, you'll need to also add that to the array you pass to `setCategories`.

**Reply by crs1138** [reply]
> 2\. Parse it into a general block data structure and compare that data to a snapshot (`wp.blockSerializationDefaultParser.parse`)

Is this [the part of the code](https://github.com/WordPress/gutenberg/blob/d8ba3977c2da51236bf817eb42f36a077364bc1b/test/integration/full-content/full-content.test.js#L135-L181) in `full-content.test.js` that is responsible for parsing the HTML into a general block data structure? The one you mentioned that might not be necessary for my use case…

My _fixtures:generate_ test fails doing this because changes to the `clientId` value and I can't tell if I need to compare this JSON or not. :) However, the rest of the test suite passes ok without this.

<img width="1061" alt="image" src="https://user-images.githubusercontent.com/1313681/166913905-b25a3c19-b971-4800-a09b-64f1e5ef236f.png">

I get the comparison of the HTML fixtures. I must admit I am a bit lost when it comes to saving the JSON files. When would the comparisons of these fail?

**Reply by crs1138** [reply]
Also, what can I do to have the clientId match?

**Reply by talldan** [reply]
The Gutenberg tests use this `normalizeParsedBlocks` function to only output a subset of the parsed block data:
https://github.com/WordPress/gutenberg/blob/d8ba3977c2da51236bf817eb42f36a077364bc1b/test/integration/full-content/full-content.test.js#L47-L53

So you could do something like that before calling `expect`.

**Reply by crs1138** [reply]
I rewrote the whole file of _full-content.test.js_ and only added a function to register our custom blocks. So, I do use the `normalizeParsedBlocks()` too. However, the `clientId`s differ. I have no idea why and to be honest I don't think I need to be testing them for our purposes. Or do I?

EDIT: Here's a [diff](https://www.diffchecker.com/HIljvt2k) so you can see my modification to the whole file.
EDIT2: Just to clarify, when I run `fixtures:regenerate` script, all tests pass ok. It's only when `fixtures:generate` runs that some of the blocks have problem with the `clientId` property.

**Reply by talldan** [reply]
I don't think you'll need to test them. `normalizeParsedBlocks` will strip them out, but you'll need to make sure you're testing the return value of `normalizeParsedBlocks`.

**Reply by crs1138** [reply]
Doh! When I was rewriting the `normalizeParsedBlocks()` I skipped the recursion to _keep it simple_ to begin with. :) And of course, I forgot all about that. Now the tests seem to pass as they supposed to. I shall extend this to all our blocks to see.

@talldan Thanks for your patience and replies. I really appreciate that.

**Reply by talldan** [reply]
No worries, glad you got it working!
